### PR TITLE
[Merged by Bors] - feat: port Logic.Equiv.TransferInstance

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Logic/Equiv/TransferInstance.lean [hunk 1]
```diff
@@ -43,131 +43,132 @@ section Instances
...
+    let mul := Equiv.Mul e
     α ≃* β := by
   intros
   exact
     { e with
       map_mul' := fun x y => by
         apply e.symm.injective
-        simp }
+        simp only [toFun_as_coe_apply, symm_apply_apply]
+        rfl }
```
Comment (eric-wieser) [inline Mathlib/Logic/Equiv/TransferInstance.lean]: It would be better to try and fix `simp` than to insert these workarounds.

Comment (Parcly-Taxel) [reply Mathlib/Logic/Equiv/TransferInstance.lean]: It's not much of a problem – just `simp; rfl` also works. Should I do that?

Comment (eric-wieser) [reply Mathlib/Logic/Equiv/TransferInstance.lean]: What you should do is look at the goal state you end up with, and try to work out which simp lemma is missing.

**Comment by joelriou** [comment]
I have succeeded in having the file compile by slightly changing the formulation of lemmas like `mul_def`:
```lean
@[to_additive]
theorem mul_def [Mul β] (x y : α) :
  letI := Equiv.Mul e
  x * y = e.symm (e x * e y) := rfl
-- instead of
--theorem mul_def [Mul β] (x y : α) : @Mul.mul _ (Equiv.Mul e) x y = e.symm (e x * e y) :=
--  rfl
```
Doing so makes `simp only [mul_def]` work (I cannot completely explain why though...). Then, I was able to make the whole file compile. However, I believe that all the lemmas `_def` in this file should be modified accordingly before this is merged.

File: Mathlib/Logic/Equiv/TransferInstance.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/Logic/Equiv/TransferInstance.lean]: The culprit was `add/mul_def`...

**Comment by eric-wieser** [comment]
> (I cannot completely explain why though...

The new spelling uses `HMul`, which is what `*` is notation for. That is why it now works.

File: Mathlib/Logic/Equiv/TransferInstance.lean [hunk 2]
```diff
@@ -43,131 +43,139 @@ section Instances
...
-    letI := Equiv.hasMul e
+    let mul := Equiv.Mul e
     α ≃* β := by
   intros
   exact
     { e with
       map_mul' := fun x y => by
         apply e.symm.injective
-        simp }
+        simp only [toFun_as_coe_apply, symm_apply_apply, mul_def] }
```
Comment (eric-wieser) [inline Mathlib/Logic/Equiv/TransferInstance.lean]: Does this `simp` need to be squeezed, or was that left from a previous fix?

**Review by eric-wieser — APPROVED** [review]

**Comment by eric-wieser** [comment]
bors merge

**Pull request closed without merging**